shared/subspace-object-fetching/src/object_fetcher.rs

```diff
@@ -0,0 +1,402 @@
+                    progress > offset_in_segment
+                })
+                .ok_or_else(|| {
+                    let error_string = format!(
+                        "Failed to find item at offset {offset_in_segment} in segment \
+                        {segment_index} for object {mapping:?}"
+                    );
+                    error!(error = %error_string);
+
+                    Error::StringError(error_string)
+                })?;
+
+            match segment_item {
+                SegmentItem::Block { bytes, .. }
+                | SegmentItem::BlockStart { bytes, .. }
+                | SegmentItem::BlockContinuation { bytes, .. } => {
+                    // Rewind back progress to the beginning of the number of bytes
+                    progress -= bytes.len();
+                    // Get a chunk of the bytes starting at the position we care about
+                    Vec::from(&bytes[offset_in_segment - progress..])
+                }
+                segment_item => {
+                    error!(
+                        ?segment_item,
+                        offset_in_segment,
+                        %segment_index,
+                        ?mapping,
+                        "Unexpected segment item",
+                    );
+
+                    return Err(Error::StringError(format!(
+                        "Unexpected segment item at offset {offset_in_segment} in segment \
+                        {segment_index} for object {mapping:?}"
+                    )));
+                }
+            }
+        };
+
+        if let Ok(data) = Vec::<u8>::decode(&mut data.as_slice()) {
+            return Ok(data);
+        }
+
+        for segment_index in segment_index + SegmentIndex::ONE.. {
+            let Segment::V0 { items } = self.read_segment(segment_index).await?;
+            for segment_item in items {
+                if let SegmentItem::BlockContinuation { bytes, .. } = segment_item {
+                    data.extend_from_slice(&bytes);
+
+                    if let Ok(data) = Vec::<u8>::decode(&mut data.as_slice()) {
+                        return Ok(data);
+                    }
+                }
+            }
+
+            if data.len() >= MAX_OBJECT_SIZE {
+                break;
+            }
+        }
+
+        error!(?mapping, "Read max object size for object without success");
+
+        Err(Error::StringError(format!(
+            "Read max object size for object without success: {mapping:?}"
+        )))
+    }
+
+    /// Read the whole segment by its index (just records, skipping witnesses)
+    // TODO: Reconstruction assumes that all source pieces are available, which is not always the case
+    async fn read_segment(&self, segment_index: SegmentIndex) -> Result<Segment, Error> {
+        let mut segment_bytes =
+            Vec::<u8>::with_capacity(ArchivedHistorySegment::NUM_PIECES * RawRecord::SIZE);
+
+        for piece_index in segment_index
```

Comment:
This code is very slow. Check `subspace-service::sync_from_dsn::import_blocks::download_and_reconstruct_blocks()` for a similar task.

Reply:
Thanks, I’m aware of that function, but I wanted to get the API and tests merged before making optimisations.

Is there a similar function that’s located in a crate which this crate can depend on? Or can we extract the logic from that function into a low-level crate?

Reply:
> Is there a similar function that’s located in a crate which this crate can depend on? Or can we extract the logic from that function into a low-level crate?

We don't have a separate crate for this function and we had a previous discussion to start adding `subspace-networking-components` crate (@nazar-pc ) but it always seemed "too small snippet to justify the new crate".

Reply:
Ok, I'll leave a TODO for now. Then in the next PR I'll copy that code, and leave comments on both copies saying where the other copy is.